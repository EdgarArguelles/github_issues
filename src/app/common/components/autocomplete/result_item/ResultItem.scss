.result-item {
  cursor: pointer;

  :hover {
    background-color: #dadada;
  }

  .content {
    padding: 15px;
  }
}