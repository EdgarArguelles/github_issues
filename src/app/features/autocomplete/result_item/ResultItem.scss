// variables
@use '../../../Variables';

.result-item {
  cursor: pointer;

  :hover {
    background-color: Variables.$hover-color;
  }

  .content {
    padding: 15px;
  }
}