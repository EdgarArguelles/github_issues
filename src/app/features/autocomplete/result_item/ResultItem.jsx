import './ResultItem.scss';
import React from 'react';
import PropTypes from 'prop-types';
import {useFocusRef} from 'app/hooks/Common';

const ResultItem = ({data, focus, onSelect}) => {
    const ref = useFocusRef(focus);
    const handleSelect = () => onSelect(data.value);
    const handleKeyDown = event => {
        if (event.keyCode === 27) {
            event.preventDefault();
            onSelect(null);
        } else if (event.keyCode === 13) {
            event.preventDefault();
            handleSelect();
        }
    };

    return (
        <div className="result-item" ref={ref} tabIndex={0} onClick={handleSelect} onKeyDown={handleKeyDown}>
            <div className="content">{data.display}</div>
        </div>
    );
};

ResultItem.propTypes = {
    data: PropTypes.object.isRequired,
    focus: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
};

export default React.memo(ResultItem);