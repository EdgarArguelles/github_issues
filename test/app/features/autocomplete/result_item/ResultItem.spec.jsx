import sinon from 'sinon';
import renderer, {act} from 'react-test-renderer';
import React from 'react';
import ResultItem from 'app/features/autocomplete/result_item/ResultItem';

describe('Components -> Autocomplete -> ResultItem', () => {
    const onSelectStub = sinon.stub();
    let component, wrapper;

    afterEach(() => onSelectStub.reset());

    const mountComponent = props => {
        wrapper = renderer.create(<ResultItem {...props}/>);
        component = wrapper.root.find(el => el.type.name === 'ResultItem');
    };

    describe('load', () => {
        it('should have required props', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            expect(component.props.data).toStrictEqual({display: 'display', value: 'value'});
            expect(component.props.focus).toStrictEqual(false);
            expect(component.props.onSelect).toStrictEqual(onSelectStub);
        });
    });

    describe('onSelect', () => {
        it('should not call onSelect when key 35', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            act(() => wrapper.root.findByType('div').props.onKeyDown({keyCode: 35, preventDefault: () => {}}));

            sinon.assert.callCount(onSelectStub, 0);
        });

        it('should call onSelect when key 27', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            act(() => wrapper.root.findByType('div').props.onKeyDown({keyCode: 27, preventDefault: () => {}}));

            sinon.assert.callCount(onSelectStub, 1);
            sinon.assert.calledWithExactly(onSelectStub, null);
        });

        it('should call onSelect when key 13', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            act(() => wrapper.root.findByType('div').props.onKeyDown({keyCode: 13, preventDefault: () => {}}));

            sinon.assert.callCount(onSelectStub, 1);
            sinon.assert.calledWithExactly(onSelectStub, 'value');
        });

        it('should call onSelect when click', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            act(() => wrapper.root.findByType('div').props.onClick());

            sinon.assert.callCount(onSelectStub, 1);
            sinon.assert.calledWithExactly(onSelectStub, 'value');
        });
    });

    describe('snapshot', () => {
        it('should render ResultItem with props', () => {
            mountComponent({
                data: {display: 'display', value: 'value'},
                focus: false,
                onSelect: onSelectStub,
            });

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });
});