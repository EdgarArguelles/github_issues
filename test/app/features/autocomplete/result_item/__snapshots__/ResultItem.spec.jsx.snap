// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components -> Autocomplete -> ResultItem snapshot should render ResultItem with props 1`] = `
<div
  className="result-item"
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="content"
  >
    display
  </div>
</div>
`;
